### PR TITLE
Initialize Gradient Checker

### LINKED ISSUE #3170: adding `get_numeric_gradient` method in Python for unit test
This is the first step of #3169.

The `get_numeric_gradient`'s algorithm is describe in http://deeplearning.stanford.edu/wiki/index.php/Gradient_checking_and_advanced_optimization .

### PR DESCRIPTION
**Pull request by reyoung**
* Add get_numeric_gradient API and its unittest.
* Fix #3170

### CODE REVIEW

File: python/paddle/v2/framework/tests/gradient_checker.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        if local_scope.find_var(output) is None:
+            local_scope.new_var(output).get_tensor()
+
+    op.infer_shape(local_scope)
+
+    # allocate output memory
+    for output in op.outputs():
+        local_scope.find_var(output).get_tensor().alloc_float()
+
+    # TODO(yuyang18): Only CPU is support now.
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/gradient_checker.py]: Maybe we only need to support CPU? And we must check the GPU forward/backward implementation is same with CPU.

File: python/paddle/v2/framework/tests/gradient_checker.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    cpu_ctx = core.DeviceContext.cpu_context()
+
+    def get_output():
+        op.run(local_scope, cpu_ctx)
+        return numpy.array(local_scope.find_var(output_name).get_tensor()).sum()
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/gradient_checker.py]: How about the op has multi-outputs？

File: paddle/pybind/pybind.cc [hunk 3]
```diff
@@ -78,8 +78,14 @@ PYBIND11_PLUGIN(core) {
...
       .def("set", paddle::pybind::PyTensorSetFromArray<int>)
-      .def("shape",
-           [](pd::Tensor& self) { return pd::vectorize(self.dims()); });
+      .def("shape", [](pd::Tensor& self) { return pd::vectorize(self.dims()); })
+      .def("set_float_element",
+           [](pd::Tensor& self, size_t offset, float f) {
+             self.data<float>()[offset] = f;
+           })
+      .def("get_float_element", [](pd::Tensor& self, size_t offset) -> float {
+        return self.data<float>()[offset];
```
Comment (qingqing01) [inline paddle/pybind/pybind.cc]: The `set_float_element` and `get_float_element` only support CPU, add comments.

File: python/paddle/v2/framework/tests/gradient_checker.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    class GetNumericGradientTest(unittest.TestCase):
+        def test_add_op(self):
+            add_op = op_creations.add_two(X="X", Y="Y", Out="Z")
+            x = numpy.random.random((10, 1)).astype("float32")
+            y = numpy.random.random((10, 1)).astype("float32")
+
+            arr = get_numeric_gradient(add_op, {'X': x, "Y": y}, 'Z', 'X')
+
+            self.assertAlmostEqual(arr.mean(), 1.0, delta=1e-2)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/gradient_checker.py]: A common check function is needed.

Comment (reyoung) [inline python/paddle/v2/framework/tests/gradient_checker.py]: `GradientChecker` framework has two steps. This PR only for calculating numeric gradient for operator.

File: paddle/pybind/pybind.cc (hunk 3 above)
Comment (reyoung) [inline paddle/pybind/pybind.cc]: Done.

File: python/paddle/v2/framework/tests/gradient_checker.py (hunk 2 above)
Comment (reyoung) [inline python/paddle/v2/framework/tests/gradient_checker.py]: It does not matter if the network has multiple outputs. Just get the `output_name`'s tensor can calculate the numeric gradient.

File: python/paddle/v2/framework/tests/gradient_checker.py (hunk 1 above)
Comment (reyoung) [inline python/paddle/v2/framework/tests/gradient_checker.py]: This because some PR is under merging

**Pull request merged**